packages/ws-server/src/webSocketServer.ts

```diff
@@ -63,10 +68,53 @@ function getMaxConnectionTTL() {
     return parseInt(process.env.WS_MAX_CONNECTION_TTL || '300000');
 }
 
-app.ws.use((ctx) => {
+// function to get maximum allowed connections from a single IP
+function getLimitPerIp() {
+    return parseInt(process.env.CONNECTION_LIMIT_PER_IP || '10');
+}
+
+async function handleConnectionClose(ctx) {
+    if (ctx.websocket.ipCounted) {
+        const {ip} = ctx.request;
+        clientIps[ip]--;
+        if (clientIps[ip] === 0) delete clientIps[ip];
+    }
+    connectedClients--;
+    relay.subs()?.unsubscribe(ctx.websocket);
+    ctx.websocket.terminate();
+}
+
+app.ws.use(async (ctx) => {
     ctx.websocket.id = relay.subs()?.generateId();
     logger.info(`${LOGGER_PREFIX} new connection ${ctx.websocket.id}`);
+
+    // Close event handle
+    ctx.websocket.on('close', async (code, message) => {
+        logger.info(`Closing connection ${ctx.websocket.id} | code: ${code}, message: ${message}`);
+        await handleConnectionClose(ctx);
+    });
+    const {ip} = ctx.request;
+
+    // Increment limit counters
     connectedClients = ctx.app.server._connections;
+
+    if (!clientIps[ip]) {
+        clientIps[ip] = 1;
+    }
+    else {
+        clientIps[ip]++;
+    }
+    ctx.websocket.ipCounted = true;
+
+    // Limit checks
+
+    // Limit connections from a single IP address
+    if (clientIps[ip] && clientIps[ip] > getLimitPerIp()) {
+        logger.info(`Maximum allowed connections from a single IP (${clientIps[ip]}) exceeded for address ${ip}`);
+        ctx.websocket.close(IP_LIMIT_ERROR.code, IP_LIMIT_ERROR.message);
```

Comment:
we should have metrics when this occurs

Reply:
There is an issue for adding metrics to the ws-server, with a bullet point of tracking the count of active connections by IP